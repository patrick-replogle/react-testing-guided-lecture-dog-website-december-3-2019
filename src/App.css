img {
  max-width: 250px;
  max-height: 250px;
  padding: 10px;
}
