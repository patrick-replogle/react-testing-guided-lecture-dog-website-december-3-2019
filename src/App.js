import "./App.css";
import React from "react";
import { useLocalStorage } from "./utils/input";
import { useDogImages } from "./utils/api";

function App(props) {
  // we're calling our custom hooks here, and passing the parameters we defined in input.js
  const [breed, setBreed] = useLocalStorage("breed", "husky");
  const [count, setCount] = useLocalStorage("count", 1);
  const [images, setImages] = useDogImages(breed, count);

  return (
    <>
      <h1>The Dog Website</h1>

      <select value={breed} onChange={e => setBreed(e.target.value)}>
        <option value="husky">Husky</option>
        <option value="beagle">Beagle</option>
        <option value="corgi">Corgi</option>
        <option value="boxer">Boxer</option>
      </select>

      <input
        type="number"
        placeholder="Count"
        value={count}
        onChange={e => setCount(e.target.value)}
      />
      <button onClick={() => setImages([])}>Clear Images</button>

      <div>
        {images.map((image, index) => (
          <img key={index} src={image} alt="Dog" />
        ))}
      </div>
    </>
  );
}

export default App;
