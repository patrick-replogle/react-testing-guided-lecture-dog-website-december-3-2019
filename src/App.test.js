import React from "react";
import axios from "axios";
// no default export, so we're importing everyting from this library
import * as rtl from "@testing-library/react";
// not importing to a variable, since this just overrides jest global variables
import "@testing-library/jest-dom/extend-expect";
import App from "./App";

//mock api request when testing axios to account for snapshot not matching due to random returns of data
jest.mock("axios", () => {
  return {
    //jest.fn acts as spy
    get: jest.fn(() =>
      Promise.resolve({
        data: {
          message: ["foo.jpg", "bar.jpg"]
        }
      })
    )
  };
});

test("Made API call", async () => {
  const wrapper = rtl.render(<App />);
  await wrapper.findAllByAltText(/dog/i);
  //spy test
  expect(axios.get).toHaveBeenCalled();
});

test("Render the heading", async () => {
  // render our React app into an in-memory DOM so we can test against it
  const wrapper = rtl.render(<App />);
  await wrapper.findAllByAltText(/dog/i);
  // element is now our dom element that we can test against
  const element = wrapper.getByText(/the dog website/i);

  // test will fail if element is not visible to our robot eyes
  expect(element).toBeVisible();
});

test("Render count input", async () => {
  const wrapper = rtl.render(<App />);
  await wrapper.findAllByAltText(/dog/i);

  // using a regular expression instead of a string allows our
  // query to be much more flexible. for example, if the text becomes
  // all uppercase, we don't want our test to break
  const element = wrapper.getByPlaceholderText(/count/i);
  expect(element).toHaveValue(1);
});

//basic snapshot test-- run 'yarn test -u' to update snapshot
test("<App /> snapshot", async () => {
  const wrapper = rtl.render(<App />);
  await wrapper.findAllByAltText(/dog/i);

  expect(wrapper.asFragment()).toMatchSnapshot();
});

//test clear button
test("test clearing of images click", async () => {
  const wrapper = rtl.render(<App />);
  await wrapper.findAllByAltText(/dog/i);

  const clearBtn = wrapper.getByText(/clear images/i);

  rtl.act(() => {
    rtl.fireEvent.click(clearBtn);
  });

  expect(wrapper.queryByAltText(/dog/i)).toBeNull();
});
