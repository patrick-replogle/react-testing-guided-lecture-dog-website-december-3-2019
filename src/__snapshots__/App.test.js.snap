// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> snapshot 1`] = `
<DocumentFragment>
  <h1>
    The Dog Website
  </h1>
  <select>
    <option
      value="husky"
    >
      Husky
    </option>
    <option
      value="beagle"
    >
      Beagle
    </option>
    <option
      value="corgi"
    >
      Corgi
    </option>
    <option
      value="boxer"
    >
      Boxer
    </option>
  </select>
  <input
    placeholder="Count"
    type="number"
    value="1"
  />
  <button>
    Clear Images
  </button>
  <div>
    <img
      alt="Dog"
      src="foo.jpg"
    />
    <img
      alt="Dog"
      src="bar.jpg"
    />
  </div>
</DocumentFragment>
`;
